### PR TITLE
feat(cubestore): Storing aggregating index

### PR DESCRIPTION
**Pull request by waralex**
**Check List**
- [ ] Tests has been run in packages where changes made if available
- [ ] Linter has been run for changed code
- [ ] Tests for the changes have been added if not covered yet
- [ ] Docs have been added / updated if required

**Description of Changes Made (if issue reference is not provided)**

This is part of the aggregating index implementation related to pre-aggregation during import and compaction

### CODE REVIEW

File: rust/cubestore/cubestore/src/store/mod.rs [hunk 1]
```diff
@@ -706,6 +712,118 @@ mod tests {
...
+                    None,
+                    Some(vec![("sum".to_string(), "sum_int".to_string())]),
+                    None,
+                )
+                .await
+                .unwrap();
+
+            let data: Vec<ArrayRef> = vec![foos, boos, sums];
+
+            let indicies = meta_store.get_table_indexes(table.get_id()).await.unwrap();
```
Comment (cristipp) [inline rust/cubestore/cubestore/src/store/mod.rs:785]: Typo: indices instead of indicies.

File: rust/cubestore/cubestore/src/store/mod.rs [hunk 2]
```diff
@@ -706,6 +712,118 @@ mod tests {
...
+
+            let aggr_index = indicies
```
Comment (cristipp) [inline rust/cubestore/cubestore/src/store/mod.rs:787]: Nit: This feels overkill. Perhaps `let aggr_index = indices[0]` or something similar?

File: rust/cubestore/cubestore/src/store/mod.rs [hunk 3]
```diff
@@ -780,6 +899,67 @@ impl ChunkStore {
         Ok(new_chunks)
     }
 
+    async fn post_process_columns(
```
Comment (cristipp) [inline rust/cubestore/cubestore/src/store/mod.rs:902]: What does `post_process_columns` function do? Can we please add a small docstring?

**Review by cristipp — COMMENTED** [review]
Sprinkled in a few notes. Looks clean, biggest hurdle is the lack of docstrings to explain a bit the intent in plain English.

File: rust/cubestore/cubestore/src/store/compaction.rs [hunk 4]
```diff
@@ -848,7 +855,31 @@ async fn merge_chunks(
...
+        res = Arc::new(HashAggregateExec::try_new(
+            AggregateStrategy::InplaceSorted,
+            Some(output_sort_order),
+            AggregateMode::Final,
+            groups,
+            aggregates,
+            res.clone(),
+            schema,
+        )?);
+    } else if let Some(key_columns) = unique_key_columns {
```
[the remark this answers is not in the document]
Comment (waralex) [reply rust/cubestore/cubestore/src/store/compaction.rs:882]: If `unique_key_columns` exists then columns of the aggregating index must be a subset of `unique_key_columns` (checked when creating the index). So aggregate_columns and unique_key_columns mutually aren't mutually exclusive, but for an aggregating index (in this context, if aggregate_columns is not None), uniqueness is provided by aggregation

File: rust/cubestore/cubestore/src/store/mod.rs (hunk 2 above)
Comment (waralex) [reply rust/cubestore/cubestore/src/store/mod.rs:787]: The fact is that with create_table, the `default` index is automatically created and perhaps some other automatic indexes will be created in the future. The position of the custom index in this list may change with minor edits in the `create_table` implementation and I would not like this to break the tests.

File: rust/cubestore/cubestore/src/metastore/table.rs [hunk 5]
```diff
@@ -1,20 +1,100 @@
...
+    pub fn aggregate_expr(
+        &self,
+        schema: &ArrowSchema,
+    ) -> Result<Arc<dyn AggregateExpr>, CubeError> {
+        let col = Arc::new(FusionColumn::new_with_schema(
+            self.column.get_name().as_str(),
+            &schema,
+        )?);
+        let res: Arc<dyn AggregateExpr> = match self.function {
+            AggregateFunction::SUM => {
```
[the remark this answers is not in the document]
Comment (waralex) [reply rust/cubestore/cubestore/src/metastore/table.rs:77]: I plan to do this in the next PR

File: rust/cubestore/cubestore/src/store/mod.rs (hunk 3 above)
Comment (waralex) [reply rust/cubestore/cubestore/src/store/mod.rs:905]: done at 8cf2c0a

File: rust/cubestore/cubestore/src/store/mod.rs (hunk 1 above)
Comment (waralex) [reply rust/cubestore/cubestore/src/store/mod.rs]: done at 8cf2c0a

File: rust/cubestore/cubestore/src/store/compaction.rs [hunk 6]
```diff
@@ -832,6 +838,7 @@ async fn merge_chunks(
     l: Arc<dyn ExecutionPlan>,
     r: Vec<ArrayRef>,
     unique_key_columns: Option<Vec<&crate::metastore::Column>>,
+    aggregate_columns: Option<Vec<AggregateColumn>>,
```
[the remark this answers is not in the document]
Comment (waralex) [reply rust/cubestore/cubestore/src/store/compaction.rs:842]: done at 8cf2c0a

File: rust/cubestore/cubestore/src/metastore/mod.rs [hunk 7]
```diff
@@ -2848,7 +2848,8 @@ impl RocksMetaStore {
         }
 
         let aggregate_columns = table_id.get_row().aggregate_columns();
-        if aggregate_columns.is_none() {
+        if aggregate_columns.is_empty() {
+            //TODO check that index column is not in aggregate columns
```
Comment (cristipp) [inline rust/cubestore/cubestore/src/metastore/mod.rs:2852]: In the long run, it is harder to fix TODOs out of band than fixing them right now. It comes down to the cost of establishing the proper brain context: now you have uploaded all this stuff in your brain and ready to go, 3 months later the lane you (or someone else) might be a bit fuzzy on details, and would be harder for you to fix the TODO.

Barring that, consider "// TODO(waralex) Check ..."

File: rust/cubestore/cubestore/src/queryplanner/info_schema/system_tables.rs [hunk 8]
```diff
@@ -144,40 +144,10 @@ impl InfoSchemaTableDef for SystemTablesTableDef {
...
-                        .map(|row| {
-                            row.table
-                                .get_row()
-                                .aggregate_columns()
-                                .as_ref()
-                                .map(|v| format!("{:?}", v))
-                        })
-                        .collect::<Vec<_>>();
-                    Arc::new(StringArray::from(
-                        aggregates
```
Comment (cristipp) [inline rust/cubestore/cubestore/src/queryplanner/info_schema/system_tables.rs:156]: Nothing to add here other than I'm very happy to see a bunch of code deleted from a long method :)

File: rust/cubestore/cubestore/src/store/compaction.rs [hunk 9]
```diff
@@ -827,11 +833,13 @@ async fn write_to_files_by_keys(
     Ok(row_counts)
 }
 
+///Makes `SendableRecordBatchStream` with result of merging  persistent chunk `l` and in-memory chunk `r`
 async fn merge_chunks(
```
Comment (cristipp) [inline rust/cubestore/cubestore/src/store/compaction.rs:837]: I'm going to attempt to help with the English, while leaving the payload up to you.

"/// Builds a `SendableRecordBatchStream` containing the result of merging a persistent chunk `l` with an in-memory chunk `r`.

Comment (cristipp) [reply rust/cubestore/cubestore/src/store/compaction.rs:837]: BTW, this is exactly the kind of 'brief docstring' I'm looking for, to give some orientation to the casual reader.

File: rust/cubestore/cubestore/src/store/compaction.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cristipp) [reply rust/cubestore/cubestore/src/store/compaction.rs:883]: My usual suggestion in such cases is to add the review response as a comment in the code. This is entirely non-obvious for people not intimately familiar with the code. Think of us 2 months ago, struggling to get onboarded with little guidance :)
